File: backend/src/routes/query.ts
import { Router, Request, Response } from 'express';
import { RAGPipeline } from '@/lib/rag-pipeline';
import { QueryResponse } from '@/types';

const router = Router();
const ragPipeline = new RAGPipeline();

interface QueryRequest {
  question: string;
}

router.post('/', async (req: Request<{}, QueryResponse, QueryRequest>, res: Response<QueryResponse>) => {
  try {
    const { question } = req.body;

    if (!question) {
      return res.status(400).json({
        success: false,
        error: 'Question is required and must be a string'
      });
    }

    const response = await ragPipeline.query(question);

    res.json({
      success: true,
      answer: response.answer,
      sources: response.sources,
      confidence: response.confidence,
    });

  } catch (error) {
    console.error('Query error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to process query'
    });
  }
});

export default router;

File: backend/internal/handlers/query.go
package handlers

import (
	"context"
	"net/http"
	"rag-backend/pkg/codes"

	"github.com/gin-gonic/gin"

	"rag-backend/internal/services"
	"rag-backend/pkg/types"
)

type QueryService interface {
	Query(question string) (*types.RAGResponse, error)
	QueryStream(ctx context.Context, question string) (<-chan services.StreamEvent, error)
}

type QueryHandler struct {
	ragPipeline QueryService
}

func NewQueryHandler(ragPipeline QueryService) *QueryHandler {
	return &QueryHandler{
		ragPipeline: ragPipeline,
	}
}

func (h *QueryHandler) HandleQuery(c *gin.Context) {
	var request types.QueryRequest

	if err := c.ShouldBindJSON(&request); err != nil {
		c.JSON(http.StatusBadRequest, types.ErrorResponse{
			Error: "Question is required and must be a string",
			Code:  codes.ErrInvalidRequest,
		})
		return
	}

	if request.Question == "" {
		c.JSON(http.StatusBadRequest, types.ErrorResponse{
			Error: "Question cannot be empty",
			Code:  codes.ErrEmptyQuestion,
		})
		return
	}

	response, err := h.ragPipeline.Query(request.Question)
	if err != nil {
		c.JSON(http.StatusInternalServerError, types.ErrorResponse{
			Error:   "Failed to process query",
			Code:    codes.ErrQueryError,
			Details: err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, types.QueryResponse{
		Answer:     response.Answer,
		Sources:    response.Sources,
		Confidence: response.Confidence,
	})
}
